src/lava/utils/float2fixed.py

```diff
@@ -1,26 +1,719 @@
+        mapping. The function must be consistent between Variables of one scale
+        domain of be chosen such that for every Variable the numerical
+        fixed-point representation can be represented with the given
+        constraints. The optimal scaling function will be stored in a
+        dictionary with keys scale domains.
+
+        Returns
+        -------
+        scaling_funct_dict: dict
+            Dictionary containing the optimal scaling function for each scale
+            domain.
+        """
+        # Initialize scaling function dictionary.
+        scaling_factors = {}
+        slopes = []
+
+        for scale_domain, scale_data_domain in self.scale_domains.items():
+            scaling_factors[scale_domain] = {}
+            for p_id, scale_data_p_id in scale_data_domain.items():
+                for var, scale_data in scale_data_p_id.items():
+
+                    bits = scale_data['precision_bits']
+                    impl_shift = scale_data['implicit_shift']
+                    num_bits_exp = scale_data['num_bits_exp']
+
+                    # Get range representable given bit constraints and
+                    # signedness.
+                    if scale_data['signedness'] == 's':
+                        rep_range = np.array([-1 * 2 ** (bits - 1),
+                                              2 ** (bits - 1) - 1])
+                    elif scale_data['signedness'] == 'u':
+                        rep_range = np.array([0, 2 ** bits - 1])
+
+                    # Apply implicit shift.
+                    rep_range = np.left_shift(rep_range, impl_shift)
+
+                    # Apply maximal shift due to exponent.
+                    rep_range = np.left_shift(rep_range, 2 ** num_bits_exp - 1)
+
+                    # Get range of Variables in floating point domain.
+                    domain = scale_data['domain']
+
+                    # If domain of Variable has been recorded apply quantile.
+                    if var in self.monitors[p_id].keys():
+                        lower_val = np.quantile(domain, self.quantiles[0])
+                        higher_val = np.quantile(domain, self.quantiles[1])
+                        dyn_range = np.array([lower_val, higher_val])
+                    else:
+                        dyn_range = np.array([np.min(domain), np.max(domain)])
+
+                    # Calculate slopes for linear mapping of var.
+                    # If an entry in rep_range or dyn_range is zero, inf will
+                    # be filled in.
+                    not_zero_cond = (rep_range != 0) * (dyn_range != 0)
+                    slopes_var = np.divide(rep_range, dyn_range,
+                                           out=np.inf * np.ones_like(dyn_range),
+                                           where=not_zero_cond)
+
+                    # Take absolute value. Only positive slopes are allowed.
+                    slopes_var = np.absolute(slopes_var)
```

Comment:
Is there a way how the slope may be negative? If so, there is something wrong. In that case, the absolute value of the slope would be a random number, and I would suggest that we throw an error instead. Or what do you think?

Reply:
I don't think there would be something wrong.
Imagine the following situation:  
For a variable `var` we use a signed variable with `7` bits, implying smallest and highest values are `-2**6` and `2**6-1*`.  
Given a specific parameter choice, `var` only attains positive values, say the smallest is `0.1` and the largest is `1.`  The slopes are calculated as  `-2**6 / 0.1` and `(2**6 -1) / 1`.    
The first now is negative, but this is now problem. Since we want, however, the smallest positive slope we calculate the absolute value of the slopes and then take the minimum.  
Does this address your question?